{
  "type": "lab",
  "title": "Error Handling and Exceptions",
  "description": "Although exceptions were introduced earlier, this section fills in some additional details about error checking and exception handling.",
  "meta": {
    "title": "Mastering Python Error Handling and Exceptions",
    "description": "Explore the essentials of error handling and exception management in Python programming.",
    "keywords": "try except, error handling, python, exceptions, runtime errors"
  },
  "difficulty": "Beginner",
  "time": 80,
  "hidden": false,
  "fee_type": "free",
  "show_in_tutorial": true,
  "details": {
    "steps": [
      {
        "title": "How programs fail",
        "text": "step1.md",
        "verify": [
          {
            "name": "Verify Presence of 'def' in Python History",
            "file": "verify1.sh",
            "hint": "Please follow the instructions to complete the steps.",
            "timeout": 0,
            "showstderr": true
          }
        ],
        "skills": [
          "python/build_in_functions",
          "python/comments",
          "python/for_loops",
          "python/function_definition",
          "python/numeric_types",
          "python/python_shell",
          "python/strings",
          "python/tuples",
          "python/variables_data_types"
        ]
      },
      {
        "title": "Exceptions",
        "text": "step2.md",
        "verify": [
          {
            "name": "This step has no verification at the moment",
            "file": "verify2.sh",
            "hint": "This step has no verification at the moment",
            "timeout": 0,
            "showstderr": true
          }
        ],
        "skills": [
          "python/build_in_functions",
          "python/catching_exceptions",
          "python/conditional_statements",
          "python/raising_exceptions",
          "python/sets"
        ]
      },
      {
        "title": "Exception Handling",
        "text": "step3.md",
        "verify": [
          {
            "name": "This step has no verification at the moment",
            "file": "verify3.sh",
            "hint": "This step has no verification at the moment",
            "timeout": 0,
            "showstderr": true
          }
        ],
        "skills": [
          "python/catching_exceptions",
          "python/comments",
          "python/function_definition",
          "python/raising_exceptions"
        ]
      },
      {
        "title": "Built-in Exceptions",
        "text": "step4.md",
        "verify": [
          {
            "name": "This step has no verification at the moment",
            "file": "verify4.sh",
            "hint": "This step has no verification at the moment",
            "timeout": 0,
            "showstderr": true
          }
        ],
        "skills": []
      },
      {
        "title": "Exception Values",
        "text": "step5.md",
        "verify": [
          {
            "name": "Verify Python History Contains Raise Statement",
            "file": "verify5.sh",
            "hint": "Please follow the instructions to complete the steps.",
            "timeout": 0,
            "showstderr": true
          }
        ],
        "skills": [
          "python/build_in_functions",
          "python/catching_exceptions",
          "python/comments",
          "python/raising_exceptions",
          "python/tuples"
        ]
      },
      {
        "title": "Catching Multiple Errors",
        "text": "step6.md",
        "verify": [
          {
            "name": "This step has no verification at the moment",
            "file": "verify6.sh",
            "hint": "This step has no verification at the moment",
            "timeout": 0,
            "showstderr": true
          }
        ],
        "skills": ["python/catching_exceptions", "python/tuples"]
      },
      {
        "title": "Catching All Errors",
        "text": "step7.md",
        "verify": [
          {
            "name": "Verify Presence of 'print' in Python History",
            "file": "verify7.sh",
            "hint": "Please follow the instructions to complete the steps.",
            "timeout": 0,
            "showstderr": true
          }
        ],
        "skills": [
          "python/build_in_functions",
          "python/catching_exceptions",
          "python/comments"
        ]
      },
      {
        "title": "Wrong Way to Catch Errors",
        "text": "step8.md",
        "verify": [
          {
            "name": "This step has no verification at the moment",
            "file": "verify8.sh",
            "hint": "This step has no verification at the moment",
            "timeout": 0,
            "showstderr": true
          }
        ],
        "skills": ["python/build_in_functions", "python/catching_exceptions"]
      },
      {
        "title": "Somewhat Better Approach",
        "text": "step9.md",
        "verify": [
          {
            "name": "This step has no verification at the moment",
            "file": "verify9.sh",
            "hint": "This step has no verification at the moment",
            "timeout": 0,
            "showstderr": true
          }
        ],
        "skills": [
          "python/build_in_functions",
          "python/catching_exceptions",
          "python/tuples"
        ]
      },
      {
        "title": "Reraising an Exception",
        "text": "step10.md",
        "verify": [
          {
            "name": "This step has no verification at the moment",
            "file": "verify10.sh",
            "hint": "This step has no verification at the moment",
            "timeout": 0,
            "showstderr": true
          }
        ],
        "skills": [
          "python/build_in_functions",
          "python/catching_exceptions",
          "python/raising_exceptions",
          "python/tuples"
        ]
      },
      {
        "title": "Exception Best Practices",
        "text": "step11.md",
        "verify": [
          {
            "name": "This step has no verification at the moment",
            "file": "verify11.sh",
            "hint": "This step has no verification at the moment",
            "timeout": 0,
            "showstderr": true
          }
        ],
        "skills": []
      },
      {
        "title": "finally statement",
        "text": "step12.md",
        "verify": [
          {
            "name": "This step has no verification at the moment",
            "file": "verify12.sh",
            "hint": "This step has no verification at the moment",
            "timeout": 0,
            "showstderr": true
          }
        ],
        "skills": [
          "python/catching_exceptions",
          "python/comments",
          "python/finally_block"
        ]
      },
      {
        "title": "with statement",
        "text": "step13.md",
        "verify": [
          {
            "name": "Verify Presence of 'with' in Python History",
            "file": "verify13.sh",
            "hint": "Please follow the instructions to complete the steps.",
            "timeout": 0,
            "showstderr": true
          }
        ],
        "skills": [
          "python/build_in_functions",
          "python/comments",
          "python/file_opening_closing",
          "python/with_statement"
        ]
      },
      {
        "title": "Exercise 3.8: Raising exceptions",
        "text": "step14.md",
        "verify": [
          {
            "name": "Verify Presence of RuntimeError Exception",
            "file": "verify14.sh",
            "hint": "Please follow the instructions to complete the steps.",
            "timeout": 0,
            "showstderr": true
          }
        ],
        "skills": [
          "python/booleans",
          "python/break_continue",
          "python/build_in_functions",
          "python/comments",
          "python/conditional_statements",
          "python/data_collections",
          "python/default_arguments",
          "python/file_opening_closing",
          "python/for_loops",
          "python/function_definition",
          "python/importing_modules",
          "python/iterators",
          "python/list_comprehensions",
          "python/lists",
          "python/python_shell",
          "python/raising_exceptions",
          "python/standard_libraries",
          "python/tuples",
          "python/variables_data_types",
          "python/with_statement"
        ],
        "solutions": ["fileparse_3.8_solution.py"]
      },
      {
        "title": "Exercise 3.9: Catching exceptions",
        "text": "step15.md",
        "verify": [
          {
            "name": "Verify Parsing of Python File",
            "file": "verify15.sh",
            "hint": "Please follow the instructions to complete the steps.",
            "timeout": 0,
            "showstderr": true
          }
        ],
        "skills": [
          "python/booleans",
          "python/break_continue",
          "python/build_in_functions",
          "python/catching_exceptions",
          "python/comments",
          "python/conditional_statements",
          "python/data_collections",
          "python/default_arguments",
          "python/dictionaries",
          "python/file_opening_closing",
          "python/for_loops",
          "python/function_definition",
          "python/importing_modules",
          "python/iterators",
          "python/list_comprehensions",
          "python/lists",
          "python/numeric_types",
          "python/python_shell",
          "python/raising_exceptions",
          "python/sets",
          "python/standard_libraries",
          "python/strings",
          "python/tuples",
          "python/type_conversion",
          "python/variables_data_types",
          "python/with_statement"
        ],
        "solutions": ["fileparse_3.9_solution.py"]
      },
      {
        "title": "Exercise 3.10: Silencing Errors",
        "text": "step16.md",
        "verify": [
          {
            "name": "Verify Parsing of MSFT Stock Data",
            "file": "verify16.sh",
            "hint": "Please follow the instructions to complete the steps.",
            "timeout": 0,
            "showstderr": true
          }
        ],
        "skills": [
          "python/booleans",
          "python/break_continue",
          "python/build_in_functions",
          "python/catching_exceptions",
          "python/comments",
          "python/conditional_statements",
          "python/data_collections",
          "python/dictionaries",
          "python/file_opening_closing",
          "python/for_loops",
          "python/function_definition",
          "python/importing_modules",
          "python/iterators",
          "python/list_comprehensions",
          "python/lists",
          "python/numeric_types",
          "python/python_shell",
          "python/raising_exceptions",
          "python/sets",
          "python/standard_libraries",
          "python/strings",
          "python/tuples",
          "python/type_conversion",
          "python/variables_data_types",
          "python/with_statement"
        ],
        "solutions": ["fileparse_3.10_solution.py"]
      }
    ],
    "intro": {
      "text": "intro.md",
      "background": "setup.sh"
    },
    "finish": {
      "text": "finish.md"
    },
    "assets": {
      "host01": [
        {
          "file": "missing.csv",
          "target": "/home/labex/project",
          "chmod": "ugo+rwx"
        },
        {
          "file": "prices.csv",
          "target": "/home/labex/project",
          "chmod": "ugo+rwx"
        },
        {
          "file": "missing.csv",
          "target": "/home/labex/project",
          "chmod": "ugo+rwx"
        },
        {
          "file": "fileparse_3.10.py",
          "target": "/home/labex/project",
          "chmod": "ugo+rwx"
        },
        {
          "file": "fileparse_3.9.py",
          "target": "/home/labex/project",
          "chmod": "ugo+rwx"
        },
        {
          "file": "fileparse_3.8.py",
          "target": "/home/labex/project",
          "chmod": "ugo+rwx"
        }
      ]
    }
  },
  "backend": {
    "imageid": "vnc-ubuntu:2204"
  },
  "contributors": [],
  "license": {
    "name": "CC-BY-SA-4.0",
    "url": "https://github.com/dabeaz-course/practical-python/blob/master/LICENSE.md",
    "repo": "https://github.com/dabeaz-course/practical-python"
  }
}
